rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,500 @@
+
+[writable-property]: https://www.typescriptlang.org/play/#code/PTAEFpNB3BLATApgO1sg5qARgJ0QQwGsBnCKEAWACg0AXRHAM3wGNFQB1WAG266WSgA3tVBjQ+AFyhitHGnQBuagF9q1Og2ZtQAEQT8Uw0eKky5C0AB9QyAK4BbLA2VU1VaozvIWtWAHtBAAt8ZHhuRGIuXkNkAAoAN3xuO0RpaL4EFABKaXsnBmMqcVA8WjscQSSUxAA6fFqIjFog13dqJBZufDxQCNoYHkyBdKHY1xCwiKixrPi4GLns1w7ELp72ftB4AznpfXhx6knwyIzYuJ3DpZWqEDJwWx6cfzgMCW5if2w8IlJIR6UDw0ZD0JisdgZAByz1eRRKZlk8neNnyzhwbXUILB2nYBxhOBe0HhpjyjnRmOBXh8fkCMHk9GIABV-NDYdBEslUqNeASidkSWJqql6qAALygADkQUQvH8kspq3WvS2C24fNePPV7Nc0AZkRZbMJrziao10GWWM63RViAGV3N+wQ5t1+uZrKG5suzvZlqoQA
+
+Each of these will trigger a compiler error for consumers, surfacing the error. As such, they should be easily detectable by testing infrastructure (see below under [Tooling: Detect breaking changes in types](#detect-breaking-changes-in-types)).
+
+There are several reasons why breaking changes may occur:
+
+-   The author of the addon may choose to change the API for whatever reason. This is no different than the situation today for addons which do not support TypeScript.
+
+-   The author of the addon may need to make changes to adapt to changes in Ember, for example to support Octane idioms. This is likewise identical with the situation for addons which do not support TypeScript.
+
+-   Adopting a new version of TypeScript may change the meaning of existing types. For example, in TypeScript 3.5, generic types without a specified default type changed their default value from `{}` to `unknown`. This improved type safety, but [broke many existing types][3.5-breakage], including the internal types for Glimmer and Ember.
+
+-   Adopting a new version of TypeScript may change the type definitions emitted in `.d.ts` files in backwards-incompatible ways. For example, changing to use the finalized ECMAScript spec for class fields meant that [types emitted by TypeScript 3.7 were incompatible with TypeScript 3.5 and earlier][3.7-emit-change].
+
+[3.5-breakage]: https://github.com/microsoft/TypeScript/issues/33272
+[3.7-emit-change]: https://github.com/microsoft/TypeScript/pull/33470
+
+#### Non-breaking changes to type definitions
+
+The following are *not* breaking changes:
+
+-   a function (including a class method or constructor) *accepts a less specific (“wider”) type*, for example if it previously accepted only a `boolean` but now accepts `boolean | undefined`—since all existing user code will continue working and type-checking ([playground][wider-argument])
+
+-   a function (including a class method) which *returns a more specific (“narrower”) type*, for example if it previously returned `number | undefined` and now always returns `number`—since all user code will continue working and type-checking ([playground][narrower-return])
```

Comment:
This is an interesting one to consider, because a non-breaking change to Ember could result in a breaking change to the types.

If originally this were returned
```ts
{
   a: number
   b?: string
}
```
and now a change to Ember is made such that `b` is _always present_
```ts
{
   a: number
   b: string
}
```
we have a narrower type than before. Ember maintained compatibility (all old code continues to work), but by this definition of "breaking change" we end up causing breakage to types when we accurately represent that `b` will under no circumstances be `undefined`.  Does this feel right to everyone?

Reply:
I'm not sure I'm following how the change you're describing would introduce type breakage. Could you give a code snippet that demonstrates the break?

Reply:
I suppose this would do it, if `Ember.theFunction` were to change from returning `{ a: number; b?: string }` to making `b` required.

```ts
let foo: ReturnType<Ember.theFunction> = { a: 123 };
```

Reply:
Relevant to [my comment above](https://github.com/typed-ember/ember-cli-typescript/pull/1158#discussion_r520906030v): I had misread this example, I think, and I need to add some nuance here. 🤔 I think I would agree with @dfreeman and *disagree* with what I wrote in my long comment above on further reflection. I wonder if we need special definitions for certain utilities like `ReturnType`.

Reply:
I *believe* I've addressed this fully via 8f4a876...ef61b3c. Thanks again for the feedback here!